### DISCUSSION TITLE
How would I install rust and cargo on a-Shell?

### DISCUSSION THREAD

**Discussion by Luxvao** [holzschu/a-shell#559] [Q&A] [answered]
I want to install the rust compiler and cargo on a-Shell. I'm on an iphone. Is that possible?
Thanks

**Comment by holzschu** [comment, marked as answer]
The short answer is no, that's not possible for the time being.

The long answer is: a-Shell can run two types of binaries: binaries compiled to Arm64 (native architecture of iPhones and iPads) and binaries compiled to WebAssembly. Before you can run a command, it has to be compiled to either of these formats, which means cross-compiling: compiling a command to a format that is not the format of the machine running the compiler. Depending on how the compile script was coded, this can range from "difficult" to "extremely difficult". As a rule of thumb, the fewer informations you have on how to compile the command, the harder it is going to be to adapt the compilation process.

For a programming language like Rust, you would also need to cross-compile the extension modules (the things that "cargo" installs), and change the way these extension modules are loaded by rust (because iOS places dynamic libraries in a non-standard place).

None of these are impossible, but they look big and scary, even for me (after the experience of running this process with clang, tex and python). It's on the todo list, but I need to have a long period of time available before I start.

**Reply by ghost** [reply]
I have a question because I’m confused, **doesn’t Rust have an ARM64 binary available?**

I ask because Rust runs on macOS, which is also ARM64, so I’m confused on where one would have to cross-compile.
The other points you make, and implied ones, are clear as day - them being:

+ Cross-compilation of extension modules (because they may not have ARM64 variant)
+ Extension-incompatibility (because of the file-location difference)
+ Rust potentially calling inaccessible libraries (i.e. system)

> To the last point, I’ve seen you have some other repos to solve some of these issues (i.e. ios_system)

I have no doubt that it still would be a difficult task, but I'm curious where cross-compilation would occur (for my own knowledge). Does Rust have critical-libraries that are extensions?

**Reply by holzschu** [reply]
The binary format for OSX Arm64 and iOS Arm64 are different. I cannot just take a binary created for macOS Arm64 and plug it it iOS. 
But judging from other languages, the big issue is likely going to be the installation and loading of extension modules.

**Reply by ghost** [reply]
Thank you for the response, makes perfect sense.
I assumed the binaries would be compatible considering Apple's switch to M1+ in their devices.

Agreed on that last point though, can imagine that’ll take either modifying the source or redirecting assembly-directives, either way you’ve got work cutout for you.

Good luck, and thanks for maintaining
